### ISSUE TITLE
The type of the result of a query seems to get lost when adding .throwIfNotFound() to it

### ISSUE THREAD

**Issue by cauboy** [Vincit/objection.js#2203]
For some reason the return type of a function gets lost when adding `throwIfNotFound()` to a query. Instead of narrowing down the type from `Promise<Application | undefined> ` to `Promise<Application>` in this example, it's now `any`.

<img width="851" alt="Screen Shot 2022-01-25 at 11 12 01" src="https://user-images.githubusercontent.com/5537152/150957459-1ec7c099-6973-4813-b145-d5b849e2c5e9.png">
<img width="598" alt="Screen Shot 2022-01-25 at 11 11 46" src="https://user-images.githubusercontent.com/5537152/150957464-ea8f90d9-69e1-4ee0-8fb8-d23cac8147a0.png">

**Comment by JeongJuhyeon** [comment]
Yeah, I can confirm this is a pretty big issue for us. It's related to a Typescript update (somewhere after 4.1). It doesn't happen on TS 4.0 + Objection 3.0. I'm not sure what the maintenance situation around this library is now given that there haven't really been any replies by the devs now for 3 months, maybe the Typescript issues that have shown up after 4.2ish are proving unfixable without a complete rewrite.
